### PR TITLE
Enhance s_client Output

### LINKED ISSUE #8123: s_client returned protocol version is confusing
When using OpenSSL 1.1.1 s_client, when connecting to server using TLS 1.2 it will print information like this (omitting certificate for brevity):
```
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: DH, 2048 bits
---
SSL handshake has read 3369 bytes and written 474 bytes
Verification: OK
---
New, TLSv1.2, Cipher is DHE-RSA-AES128-GCM-SHA256
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : DHE-RSA-AES128-GCM-SHA256
    Session-ID: 1A492810EBC7EED481E317C24F7603B8815412B878968E1834A68A9066A8E1E6
    Session-ID-ctx: 
    Master-Key: A4FF98985C1F686CE3D0C70D7A34A99D0096243A12BD43296FBB3D26A79372C51981059BCC015E33957C60F10C9601B3
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1548865483
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```
The version is returned in two places:
```
New, TLSv1.2, Cipher is DHE-RSA-AES128-GCM-SHA256
```
and
```
    Protocol  : TLSv1.2
```

when negotiating some other cipher and protocol, the information may look like this:
```
---
Acceptable client certificate CA names
O = Example CA
Client Certificate Types: RSA sign, ECDSA sign, DSA sign
Peer signing digest: MD5-SHA1
Peer signature type: RSA
Server Temp Key: DH, 2048 bits
---
SSL handshake has read 3426 bytes and written 1551 bytes
Verification: OK
---
New, SSLv3, Cipher is DHE-RSA-AES128-SHA
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.1
    Cipher    : DHE-RSA-AES128-SHA
    Session-ID: 692C1185B5CB799A4E2689F3852B0DADF8C250B2123E8EBC7892D08D2EBE94DF
    Session-ID-ctx: 
    Master-Key: C67509476284C9D75CEFC9D0FF68DDBAAA74C61A239255A3599FF4CB15FE29F5F356E19000B9487C7E1E7F12259FE5D1
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1548865404
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```
Giving two *different* protocol versions:
```
New, SSLv3, Cipher is DHE-RSA-AES128-SHA
```
and
```
    Protocol  : TLSv1.1
```

why is that? the version returned in the `New` line is not the version of the _connection_. it's the protocol version of the _cipher_!

The bigger problem is that when negotiating TLS 1.3 the *complete* output from s_client may look like this (when the server doesn't send NewSessionTicket):
```
CONNECTED(00000003)
depth=0 CN = localhost
verify error:num=18:self signed certificate
verify return:1
depth=0 CN = localhost
verify return:1
---
Certificate chain
 0 s:CN = localhost
   i:CN = localhost
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIC+zCCAeOgAwIBAgIJALEP97+mvF6wMA0GCSqGSIb3DQEBBQUAMBQxEjAQBgNV
BAMMCWxvY2FsaG9zdDAeFw0xNTAzMDMxNTUzNDhaFw0yNTAyMjgxNTUzNDhaMBQx
EjAQBgNVBAMMCWxvY2FsaG9zdDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoC
ggEBAOYz+2ar/b83RjKoFvYG7Zff7N0BT7qWdSzP6k03r2czbRo0+L1Afh6TLGz/
H66PftmXDaEPADmmJ8DjiQW05eTDV7HiVLXHG1YK4a7VhdMODFLUjV1w1rip1i/C
WAEbLHZPQZ0gN4bexY1Y7HY5cF5brCbRFHGpSHIj1Jdo3zJBawJpqM9nHDz04nVd
pJKFat8PbGrrGZBhXpg/opwg5GVAbbVroG8Nt7ATeIzY4YCaUS/9rs7J0qV6tatz
egUEGg6aCdsu4FrGQFA2/RRCtEF8H5A6gPdwTex4COR+RVXbaWUnlwyfTRjAVF2x
FiWA0WShLeTblixMxv42qUZidHUCAwEAAaNQME4wHQYDVR0OBBYEFFpCTwDYmMcu
zo+fxFRZJTIChcHqMB8GA1UdIwQYMBaAFFpCTwDYmMcuzo+fxFRZJTIChcHqMAwG
A1UdEwQFMAMBAf8wDQYJKoZIhvcNAQEFBQADggEBABF66SE6d9Px5nTjuwvq3112
Vw3l1ejJFwauYY36ctPnuNgG0V0A7Hs5ksAvsSe+SGedzASr9vRDy/aZ007Voo4a
tmP3vVtyjT2EHSXWFxLRNVr+vPnFCEBjPVpbmRNeB5odGjFh9oG6UmcdvlS/4rAI
pVOeGw0kx/yXpGRkVB6kc+XcsEi45ERIEYBBdE1RU04y/Q0YfPi6jXF6hTV7bIoH
LEKV/td6Edje2ihjlz33amR8qOv6wH1vUiRj3VRp+aegnigTq3S+FO9mmh7tP/Wg
DOOM6qxItHCcnphOYzY2gaQM5YnsLVmr3HUn/IEUT4uf2VjDNz/a7JWebscQsZI=
-----END CERTIFICATE-----
subject=CN = localhost

issuer=CN = localhost

---
No client certificate CA names sent
Peer signing digest: SHA512
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 1257 bytes and written 391 bytes
Verification error: self signed certificate
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 18 (self signed certificate)
---
```

Meaning that there is no information about negotiated protocol version, only the information derived from the cipher negotiated.

So I think that the information returned by `s_client` is confusing and unreliable. Either the `New` line should provide the protocol version of the connection, or the output should include a new line that provides the negotiated protocol version, even if the server didn't send a NST.

### PR DESCRIPTION
**Pull request by erbsland-dev**
Fixes #8123: Clarify cipher and protocol version display
- Moved the cipher version output into brackets after the cipher to clearly associate it with the cipher.
- Added a new line “Protocol:” to display the protocol version separately after the cipher line.

### CODE REVIEW

File: apps/s_client.c [hunk 1]
```diff
@@ -3481,8 +3481,9 @@ static void print_stuff(BIO *bio, SSL *s, int full)
     print_verify_detail(s, bio);
     BIO_printf(bio, (SSL_session_reused(s) ? "---\nReused, " : "---\nNew, "));
     c = SSL_get_current_cipher(s);
-    BIO_printf(bio, "%s, Cipher is %s\n",
-               SSL_CIPHER_get_version(c), SSL_CIPHER_get_name(c));
+    BIO_printf(bio, "Cipher is %s (%s)\n",
+               SSL_CIPHER_get_name(c), SSL_CIPHER_get_version(c));
```
Comment (t8m) [inline apps/s_client.c:3485]: Maybe for now just keep the Cipher output as is?

Comment (erbsland-dev) [reply apps/s_client.c]: Yes, I can do that.

I'm tempted to solve it like this: 😉
```c
    if (OSSL_PROVIDER_available(NULL, "gostprov"))
        BIO_printf(bio, "%s, Cipher is %s\n",
                   SSL_CIPHER_get_version(c), SSL_CIPHER_get_name(c));
    else
        BIO_printf(bio, "Cipher is %s (%s)\n",
                   SSL_CIPHER_get_name(c), SSL_CIPHER_get_version(c));
```

Comment (t8m) [reply apps/s_client.c]: No, that would not be a solution - that would be an unacceptable hack.

File: apps/s_client.c [hunk 2]
```diff
@@ -15,6 +15,7 @@
 #include <string.h>
 #include <errno.h>
 #include <openssl/e_os2.h>
+#include <openssl/provider.h>
```
Comment (t8m) [inline apps/s_client.c:18]: This should not be needed?

Comment (erbsland-dev) [reply apps/s_client.c]: Indeed, I'll remove that. It was a leftover from my earlier joke with the provider test.

**Comment by erbsland-dev** [comment]
@t8m Adding just the additional protocol line works. However, this PR only partially fixes issue #8123. Shall I remove the "fixes" part from the commit?

**Comment by t8m** [comment]
> @t8m Adding just the additional protocol line works. However, this PR only partially fixes issue #8123. Shall I remove the "fixes" part from the commit?

I do not think we will be doing further fixing for #8123 any time soon.

I.e., I can envision some rework of the s_client output in 4.0 or 5.0 release but I am not sure keeping #8123 open for that is useful.

**Review by t8m — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Squashed and merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**